### PR TITLE
MetaMask Extension e2e tests Best practises

### PR DESCRIPTION
**Pull request by chloeYue**
MetaMask Extension e2e tests Best practises doc

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/1999

### CODE REVIEW

File: docs/extension-e2e-test-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+- Only test things that you can control
+- Explicitly state assertions
+- Control the extension state without going through the UI
+- Avoid relying on the completion of another test in order for a test to be successful
+- Write repeatable tests that produce the same result every time
+- Keep testing modules up to date
+
+### Key
+
+✅ Recommned ⚠️ Use with caution, see notes ❌ Avoid 🤔 Explore usage
```
Comment (davidmurdoch) [inline docs/extension-e2e-test-guidelines.md:41]:
```suggestion
✅ Recommend⚠️ Use with caution, see notes ❌ Avoid 🤔 Explore usage
```

File: docs/extension-e2e-test-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+
+This document serves as a guideline to approach end-to-end test automation in our current environment so that they provide fast feedback, become less fragile, painless to debug, scalable and easy to maintain. The remainder of this document will attempt to explain the suggested approach with relevant examples to help better understand the future direction for tests.
+
+These guidelines aren't meant to be a strict set of rules, they should be thought of as an approach to writing robust end-to-end tests. Use them where they make sense and adapt them as necessary to complement your tests.
+
+### TLDR
```
Comment (davidmurdoch) [inline docs/extension-e2e-test-guidelines.md:26]: nit:

```suggestion
### TL;DR
```

File: docs/extension-e2e-test-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+'{ text: Backup your Secret Recovery Phrase to keep your wallet and funds secure, tag: div }';
+```
+
+## Wait for commands <a name="wait-for-commands"></a>
+
+Using blocking sleep calls should be avoided, instead, we should leverage wait-for commands for synchronisation. Sleeps introduce static delays, causing a test to delay the execution for a fixed period of time. This approach is inefficient, too short and the test fails, too long results in slower execution. Wait commands wait until a specific condition is met before proceeding. They reduce unnecessary waiting and ensure the tests continue as soon as the desired state is reached, making the tests more reliable and performant. Selenium Webdriver provides 3 different waiting strategies:
+
+- Explicit pauses the execution of the test until a condition is met
+- Implicit polls the DOM for a specified period of time when waiting for elements
+- Fluent specifies the amount of time to wait for a condition, as well as the frequency to check the condition
```
Comment (davidmurdoch) [inline docs/extension-e2e-test-guidelines.md:155]: nit: The wait strategy names and their definitions could use some separation:

```suggestion
- _Explicit_ - pauses the execution of the test until a condition is met
- _Implicit_ - polls the DOM for a specified period of time when waiting for elements
- _Fluent_ - specifies the amount of time to wait for a condition, as well as the frequency to check the condition
```

File: docs/extension-e2e-test-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+1. [General](#general)
+2. [Test names](#test-names)
+3. [Organization of test files](#organization-of-test-files)
+4. [Element locators](#element-locators)
+5. [Wait for commands](#wait-for-commands)
+6. [Assertions](#assertions)
+7. [Controlling state](#controlling-state)
+8. [Enhancing test stability with request mocking](#enhancing-test-stability-with-request-mocking)
+9. [Test Atomicity and Smart Test Coupling](#test-atomicity-and-smart-test-coupling)
+10. [Page Object Model](#page-object-model)
```
Comment (davidmurdoch) [inline docs/extension-e2e-test-guidelines.md:14]: GitHub itself automatically creates a table of contents, so this section might be redundant (and easy to get out of sync in the future)?

**Review by davidmurdoch — COMMENTED** [review]
mostly minor nits, feel free to ignore

**Review by mcmire — COMMENTED** [review]
Hi @chloeYue! Thanks for putting this together. I've left some comments on wording improvements below. The guidelines make sense overall but I'll spend some more time later taking a closer look at them.

File: docs/extension-e2e-test-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+
+The process of creating an E2E test involves several steps: identification, development, testing, and maintenance. Initially, we need to identify what to test, ensuring we focus on the right aspects of the product. Next, we develop the tests, adhering to the quality guidelines outlined in this document. This includes creating clear test cases, using appropriate assertions, and structuring the tests for easy maintenance. When it comes to merging, it's crucial to ensure the robustness and reliability of the test. This means the test should consistently produce the same results and be resilient to minor system changes. By following these steps, we can create effective E2E tests that contribute to delivering a high-quality product.
+
+This document serves as a guideline to approach end-to-end test automation in our current environment so that they provide fast feedback, become less fragile, painless to debug, scalable and easy to maintain. The remainder of this document will attempt to explain the suggested approach with relevant examples to help better understand the future direction for tests.
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:22]: Yes i get your point, make sense to me, thanks Elliot !

File: docs/extension-e2e-test-guidelines.md (hunk 4 above)
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md]: I updated to use the automatically generated table of content. Thanks David !

File: docs/extension-e2e-test-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+3. [Organization of test files](#organization-of-test-files)
+4. [Element locators](#element-locators)
+5. [Wait for commands](#wait-for-commands)
+6. [Assertions](#assertions)
+7. [Controlling state](#controlling-state)
+8. [Enhancing test stability with request mocking](#enhancing-test-stability-with-request-mocking)
+9. [Test Atomicity and Smart Test Coupling](#test-atomicity-and-smart-test-coupling)
+10. [Page Object Model](#page-object-model)
+
+## General <a name="general"></a>
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md]: I updated to use the automatically generated table of content instead of hardcoded url, Thanks Elliot !

File: docs/extension-e2e-test-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+
+Remember, the goal is to create tests that are reliable, easy to understand, and provide valuable feedback about your system. Whether you choose to isolate or combine tests will depend on what you're trying to achieve within your tests. Ideally, we should aim for a large number of unit tests to test individual code pieces and a medium number of E2E tests for user flow testing that cover all the scenarios.
+
+## Page Object Model <a name="page-object-model"></a>
+
+We would like to adopt the POM pattern. Similar to Mariona’s existing proposal [Page Object Model (POM) - Proposal](https://docs.google.com/document/d/1Hi64lJ8ZvTXNaoJZeh23_7HjhkRKtlcXrnYZKtQWeWk/edit#heading=h.y7td1yioh1l3). However, in the interim, there are steps we can take to reduce duplication and have clear separations of concerns in our helper files. A few of these can be found below.
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:536]: He Elliot, yes i agree maybe we should not mention the Page Object pattern now in the doc as it has not started. Will remove this section.

File: docs/extension-e2e-test-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+
+### Key
+
+✅ Recommned ⚠️ Use with caution, see notes ❌ Avoid 🤔 Explore usage
+
+## Test names <a name="test-names"></a>
+
+The test name should communicate the purpose and behaviour of the test. A clear test name serves as a concise summary of what the test aims to verify, making it easier for everyone to understand and maintain the tests. Clear test names improve the readability of the test and help when it comes to debugging a failed test. You should be able to figure out the purpose of the test without going through the complete implementation. If the test is failing, it should be easy to figure out which functionality is broken from the test name.
+
+### Guidelines
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:34]: Guidelines are a set of instructions, recommendations, or standards that provide specific directions or suggestions on how to do something or achieve a particular goal. They offer guidance or best practices for a particular task, process, or situation.
Proposal: A proposal provides guidance and instructions that have already been implemented in the codebase. In other words, proposals address existing elements in the codebase that do not adhere to best practices and need improvement.
I agree with you that the list is more readable in md files, we have converted all the tables in this document in list.

File: docs/extension-e2e-test-guidelines.md [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+
+### Proposal
+
+We propose to reorganise tests into folders based on scenarios and features. This means that each type of scenario will have its own folder, and each feature team may own one or more folders. The reasons for this proposal are as follows:
+
+- This approach provides ownership of end-to-end (E2E) testing at the feature team level. Each feature team is aware of the features they own, making it easier for them to understand what tests we currently have and what's missing.
+- As we scale, we may reach a point where core teams can no longer keep track of each of the necessary test cases for specific team features. This knowledge would be better managed by the specific feature teams.
+- We eliminate the need to decide at a low level where to place a test, as it would be straightforward based on the feature or scenario.
+- This approach aligns with our mobile testing strategy. Our mobile team prefers to have feature team tests under the same folder. This could reduce friction when switching context.
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:87]: That makes sense. I have revised the sentences here to sound more like definitive guidelines rather than discussions.

File: docs/extension-e2e-test-guidelines.md [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+
+Proposed solution: Remove conditional statements in test, structure the tests in a way that avoids branching logic.
+
+## Assertions <a name="assertions"></a>
+
+Assertions should verify expected behaviour and outcomes during test execution. When tests are responsible for assertions rather than being hidden in helper functions, it promotes test readability as expected results are explicitly stated. This makes it easier for engineers to quickly understand the intent of the test as well as the cause of failures.
+
+### Guidelines
+
+Please review the following video to get a better understanding of the guidelines below: [Assertions proposal.mov](https://drive.google.com/file/d/1ocwm0mwqXyBYgzuDB17OVYbQ2naHz4mR/view)
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:263]: Nice catch, thank you. This video explains the race condition, which is already explained in the document. I think I will just remove the link to avoid redundancy.

File: docs/extension-e2e-test-guidelines.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:34]: After reviewing the entire document, I've combined the guidelines and proposals. I believe this is a good approach to avoid confusing readers.

**Review by brad-decker — APPROVED** [review]

**Review by davidmurdoch — DISMISSED** [review]

**Review by hjetpoluru — APPROVED** [review]

File: docs/extension-e2e-test-guidelines.md [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+To ensure test isolation and simplify parallelisation, each test is run with a dedicated browser and dedicated mock services (e.g. fixtures, metrics, etc.). We use the `withFixtures` function to create everything the test needs and to perform all cleanup steps after each test, ensuring that they are completely isolated from each other. We should avoid the use of shared mocks and services between tests.
+On the other hand, allowing a degree of test coupling in E2E tests can enhance our testing effectiveness. This approach allows us to construct tests that combine multiple features, providing a closer approximation to real user behaviour. Unlike our E2E tests, real users do not commence each task with a 'clean' setup. Combining several atomic tests into a single user-flow based test may increase the execution time for individual tests. However, it eliminates the need for repeated setup, extension installation, and opening of the extension. As a result, the total test execution time is likely to decrease.
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply docs/extension-e2e-test-guidelines.md:493]: Thanks @mcmire  Indeed, I intended to include these line breaks here. I will also check again to see if there are other instances of the word 'we' that need to be updated.

**Review by mcmire — APPROVED** [review]
Great! Glad to see more documentation getting added to this repo. I'm sure we will make more improvements as time goes on but it seems there is some good advice here and I can see how this would help.

**Comment by HowardBraham** [comment]
I think we should add a section about not putting too many tests in the same `spec` file, and making sure that the runtime of any one `spec` file is less than about 2 minutes.

The reason for this is because a `spec` file is the basic unit of splitting and retrying on CircleCI.  If there are 10 tests in a `spec` file and one of them fails, all 10 tests have to be retried.  This also applied to "Rerun failed tests."

**Comment by chloeYue** [comment]
> fd55923

Good suggestion, i added a section for that, thanks !

**Comment by chloeYue** [comment]
> I think we should add a section about not putting too many tests in the same `spec` file, and making sure that the runtime of any one `spec` file is less than about 2 minutes.
> 
> The reason for this is because a `spec` file is the basic unit of splitting and retrying on CircleCI. If there are 10 tests in a `spec` file and one of them fails, all 10 tests have to be retried. This also applies to "Rerun failed tests."
> 
> Btw, _we theoretically could_ split by individual test on CircleCI, but that takes quite significant additional effort, and it's not the way it's set up right now.

Good suggestion, i added a section for that, thanks !  please check: https://github.com/MetaMask/contributor-docs/pull/72/commits/fd55923014b3a273858001954c96b1f028ca8c79

**Review by HowardBraham — APPROVED** [review]
